### ISSUE TITLE
tools/i18n/pygettext.py encoding problem

### ISSUE THREAD

**Issue by LY1806620741** [python/cpython#110522]
# Bug report

### Bug description:

1. this is `main.py`
```python
#!/usr/bin/python
# -*- coding: UTF-8 -*-
import gettext
zh = gettext.translation('messages', localedir='locale', languages=['zh'])
zh.install()
_ = zh.gettext
if __name__=='__main__':
     print(_("Hello world"))
```
2.execute `python pygettext main.py`, then generate `messages.pot` with `<_io.TextIOWrapper name='messages.pot' mode='w' encoding='cp936'>`
```
# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR ORGANIZATION
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"POT-Creation-Date: 2023-10-09 00:18+0800\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=cp936\n"
"Content-Transfer-Encoding: 8bit\n"
"Generated-By: pygettext.py 1.5\n"


#: main.py:8
msgid "Hello world"
msgstr "中文"
```
but. this file is utf8 encode when i open it with `vscode` or `windows notepad`. so when I input zh_CN word, the problem occurred.

3. `mv message.pot message.po && msgfmt message`
4. `mv message.mo locale/zh_CN/LC_MESSAGES/`
5. `python main.py` then
```log
UnicodeDecodeError: 'gbk' codec can't decode byte 0x80 in position
```

Are there any optimization points that I can do, except for manually modifying the file encoding？replace `charset=cp936` to `charset=utf8`

### CPython versions tested on:

3.11

### Operating systems tested on:

Windows

**Comment by LY1806620741** [comment]
我的解法是在项目增加`pygettext.py`的增强脚本。My solution is to add an enhanced script for 'pygettext. py' in the project

pygettext.py
```python
#!/usr/bin/python
# -*- coding: UTF-8 -*-
import sys,os
from pathlib import Path

if sys.version_info>=(3,10):
    pythonpath = sys.executable
    pygettext=Path(pythonpath+"/../Tools\i18n\pygettext.py")
    if pygettext.exists():
        os.system(pythonpath+" -X utf8 "+str(pygettext)+" "+" ".join(sys.argv[1:]))
```

以代替原来的pygettext工具。以代替原来的pygettext工具

**Comment by 522247020** [comment]
@LY1806620741 这个问题好像还是没有解决。
我查看源代码发现 

```python
    # write the output
    if options.outfile == '-':
        fp = sys.stdout
        closep = 0
    else:
        if options.outpath:
            options.outfile = os.path.join(options.outpath, options.outfile)
        fp = open(options.outfile, 'w')  # 这里是写入文件
        closep = 1
    try:
        eater.write(fp)  # 这里调用了写入文件头标识的命令
    finally:
        if closep:
            fp.close()
```

```python
    def write(self, fp):
        options = self.__options
        timestamp = time.strftime('%Y-%m-%d %H:%M%z')
        encoding = fp.encoding if fp.encoding else 'UTF-8'  # 这里获取到写入文件的 encoding 是 cp936
        print(pot_header % {'time': timestamp, 'version': __version__,
                            'charset': encoding,
                            'encoding': '8bit'}, file=fp)
```
我也不理解，哪里获取到的 cp936 ,查看终端的编码参数

```pwsh
PS > echo $OutputEncoding

Preamble          : 
BodyName          : utf-8
EncodingName      : Unicode (UTF-8)
HeaderName        : utf-8
WebName           : utf-8
WindowsCodePage   : 1200
IsBrowserDisplay  : True
IsBrowserSave     : True
IsMailNewsDisplay : True
IsMailNewsSave    : True
IsSingleByte      : False
EncoderFallback   : System.Text.EncoderReplacementFallback
DecoderFallback   : System.Text.DecoderReplacementFallback
IsReadOnly        : True
CodePage          : 65001
```
--- 

或者手动改变 `pygettext.py` 的代码

在 670 行改动

```python
fp = open(options.outfile, 'w', encoding='utf-8')
```

**Comment by 522247020** [comment]
我想问下对于 f 字符串的解决方法是什么？

```python
import gettext

_ = gettext.gettext
print(_('hello world'))

name = "Python"
print(_(f'hello {name}'))
```
对上面的代码执行 

`python ...\pygettext.py .\test.py` 会出现下面的错误:

`*** .\test.py:8: Seen unexpected token "f'"`

目前 f 字符串的使用频率非常高，不知道后面会不会支持 f 字符串

**Comment by LY1806620741** [comment]
> @LY1806620741 这个问题好像还是没有解决。 我查看源代码发现
> 
> ```python
>     # write the output
>     if options.outfile == '-':
>         fp = sys.stdout
>         closep = 0
>     else:
>         if options.outpath:
>             options.outfile = os.path.join(options.outpath, options.outfile)
>         fp = open(options.outfile, 'w')  # 这里是写入文件
>         closep = 1
>     try:
>         eater.write(fp)  # 这里调用了写入文件头标识的命令
>     finally:
>         if closep:
>             fp.close()
> ```
> 
> ```python
>     def write(self, fp):
>         options = self.__options
>         timestamp = time.strftime('%Y-%m-%d %H:%M%z')
>         encoding = fp.encoding if fp.encoding else 'UTF-8'  # 这里获取到写入文件的 encoding 是 cp936
>         print(pot_header % {'time': timestamp, 'version': __version__,
>                             'charset': encoding,
>                             'encoding': '8bit'}, file=fp)
> ```
> 
> 我也不理解，哪里获取到的 cp936 ,查看终端的编码参数
> 
> ```powershell
> PS > echo $OutputEncoding
> 
> Preamble          : 
> BodyName          : utf-8
> EncodingName      : Unicode (UTF-8)
> HeaderName        : utf-8
> WebName           : utf-8
> WindowsCodePage   : 1200
> IsBrowserDisplay  : True
> IsBrowserSave     : True
> IsMailNewsDisplay : True
> IsMailNewsSave    : True
> IsSingleByte      : False
> EncoderFallback   : System.Text.EncoderReplacementFallback
> DecoderFallback   : System.Text.DecoderReplacementFallback
> IsReadOnly        : True
> CodePage          : 65001
> ```
> 
> 或者手动改变 `pygettext.py` 的代码
> 
> 在 670 行改动
> 
> ```python
> fp = open(options.outfile, 'w', encoding='utf-8')
> ```

这是因为.Because.

https://github.com/python/cpython/blob/a9a74da4a0ca0645f049e67b6434a95e30592c32/Tools/i18n/pygettext.py#L667

创建了这个文件，用的是默认的系统编码gbk(cp936).This file was created using the default system encoding gbk (cp936)
https://github.com/python/cpython/blob/a9a74da4a0ca0645f049e67b6434a95e30592c32/Lib/_pyio.py#L74
https://github.com/python/cpython/blob/a9a74da4a0ca0645f049e67b6434a95e30592c32/Lib/_pyio.py#L59
https://github.com/python/cpython/blob/a9a74da4a0ca0645f049e67b6434a95e30592c32/Lib/locale.py#L619 尝试获取不同系统的编码.Try to get the encoding for different systems
https://github.com/python/cpython/blob/a9a74da4a0ca0645f049e67b6434a95e30592c32/Python/fileutils.c#L914

在写入文件，组装文件注释段的时候，读取了这个文件的编码，因此是cp936.When writing the file and assembling the file comment section, the encoding of the file is read, so it is cp936
https://github.com/python/cpython/blob/a9a74da4a0ca0645f049e67b6434a95e30592c32/Tools/i18n/pygettext.py#L465

**Comment by 522247020** [comment]
谢谢。

**Comment by LY1806620741** [comment]
中文：
后来我仔细看过了代码，确认问题是文本编辑器无法识别文本是gbk编码还是utf8编码引起的。非pygettext本身的错误，不知道处于什么设计思考，使用了本地编码而不是国际化编码。

如果你 使用 `python -m Tools.i18n.pygettext` 得到的是你地区的编码文件，在中国他是gbk，这时候你用现代编辑器比如`vscode`打开这个文件他将会默认识别成为utf8（然而window 的notepad都能正确识别utf8为什么不直接默认utf8..），如果你的源第一语言非通用编码就会有乱码，这时候最好是修改编码为标记编码，这样就不会有异常问题，如果你输入了地方语言并保存，那就会有不一致的问题。

总之，以标注的编码为准，如果编辑器识别错误，修改他为正确的编码再编辑。

或者方法二：强制指定翻译文件为utf8,管他什么地方编码,使用我的加强本
https://github.com/python/cpython/issues/110522#issuecomment-1752101208
或者手动指定基本编码..
`python -X utf8 -m Tools.i18n.pygettext .`

english:
Later, I looked at the code carefully and confirmed that the problem was caused by the text editor's inability to recognize whether the text was gbk encoded or utf8 encoded. It is not an error of pygettext itself. I don’t know what the design thinking is. I used local encoding instead of international encoding.
If you use `python -m Tools.i18n.pygettext` to get the encoding file in your region, which is gbk in China, then if you use a modern editor such as `vscode` to open the file, it will be recognized as utf8 by default ( However, the notepad of window can correctly recognize utf8, why not directly default to utf8..). If your source language is not a universal encoding, there will be garbled characters. At this time, it is best to change the encoding to mark encoding, so that there will be no abnormalities. The problem is, if you enter the local language and save it, there will be inconsistencies.
In short, the marked code shall prevail. If the editor recognizes it incorrectly, modify it to the correct code and then edit.
Or method two: force the translation file to be utf8, no matter where it is encoded, use my enhanced version
https://github.com/python/cpython/issues/110522#issuecomment-1752101208
Or manually specify the basic encoding..
`python -X utf8 -m Tools.i18n.pygettext .`

**Comment by LY1806620741** [comment]
> ```python
> gettext
> ```

这是不安全的，因为f会去读取变量标志，这意味着翻译文件对代码有入侵，即影响代码运行，可以查看各种变量。

如果你对`f"hello {word}"`的word有翻译需求，那么拆分开他们 `f'{_("hello")} {_("word")}'`,如果没有拆分翻译的需求,合并起来

**Comment by 522247020** [comment]
这里有个有趣的地方，我修改了我的系统语言

![image](https://github.com/python/cpython/assets/38172233/bbe855f4-e820-438f-9f4e-0b3b7dd7aca0)

这时候取到的 `charset` 是 `cp65001`

```po
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=cp65001\n"
"Content-Transfer-Encoding: 8bit\n"
"Generated-By: pygettext.py 1.5\n"
```
这时候执行 `msgfmt.py` 和运行代码都是没有问题的。
可以正常的使用！

**Comment by LY1806620741** [comment]
> ```po
> cp65001
> ```

hhh,cp65001就是utf8
https://learn.microsoft.com/en-us/answers/questions/944296/problem-with-cp-65001-(utf-8)-in-textwindow

这种方式可能会给系统带来不稳定，windows默认utf8编码是个实验版本。

**Comment by LY1806620741** [comment]
> 这里有个有趣的地方，我修改了我的系统语言
> 
> ![image](https://private-user-images.githubusercontent.com/38172233/334320212-bbe855f4-e820-438f-9f4e-0b3b7dd7aca0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY4ODU4MDMsIm5iZiI6MTcxNjg4NTUwMywicGF0aCI6Ii8zODE3MjIzMy8zMzQzMjAyMTItYmJlODU1ZjQtZTgyMC00MzhmLTlmNGUtMGIzYjdkZDdhY2EwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI4VDA4MzgyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhOWUyMTU3YzI0MWI0ZjQwNWZiMmUzODZiOWI0YzcwZGNjNDkxNzU0N2I5ZmVmYTQ1MzE0ZDIzZTkxM2E3MTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2sor68YK6PweTnoq9kM69vCjpA-3rmX-NOPulWAnlKI)
> 
> 这时候取到的 `charset` 是 `cp65001`
> 
> ```po
> "MIME-Version: 1.0\n"
> "Content-Type: text/plain; charset=cp65001\n"
> "Content-Transfer-Encoding: 8bit\n"
> "Generated-By: pygettext.py 1.5\n"
> ```
> 
> 这时候执行 `msgfmt.py` 和运行代码都是没有问题的。 可以正常的使用！


hhh,cp65001就是utf8
https://learn.microsoft.com/en-us/answers/questions/944296/problem-with-cp-65001-(utf-8)-in-textwindow

这种方式可能会给系统带来不稳定，windows默认utf8编码这个功能是个实验版本。
